tests/h/services/group_list_test.py

```diff
@@ -28,6 +28,70 @@ def test_it_includes_user_groups(self, svc, user, default_authority, sample_grou
         assert sample_groups["private"] in groups
 
 
+class TestAssociatedGroups(object):
+    def test_it_does_not_return_world_group(self, svc, sample_groups, user):
+        groups = svc.associated_groups(user)
+
+        assert "__world__" not in [group.pubid for group in groups]
+
+    def test_it_returns_user_private_groups(self, svc, sample_groups, user):
+        groups = svc.associated_groups(user)
+
+        assert sample_groups["private"] in groups
+
+    def test_it_returns_restricted_groups_if_user_is_member(self, svc, factories, user):
+        restricted_group = factories.RestrictedGroup(
+            members=[user], authority=user.authority
+        )
+        groups = svc.associated_groups(user)
+
+        assert restricted_group in groups
+
+    def test_it_returns_public_groups_if_user_is_creator(
+        self, svc, sample_groups, user
+    ):
+        groups = svc.associated_groups(user)
+
+        # Both the open and restricted groups are in the results even though the
+        # user is not a member—the user is the creator, so they are included
+        assert sample_groups["open"] in groups
+        assert sample_groups["restricted"] in groups
+
+    def test_it_does_not_return_private_groups_if_user_is_not_member(
+        self, svc, factories, user
+    ):
+        # This one's a little interesting. Recall that a user may "leave" a private
+        # group. However, that group is not deleted in this case, and they are still
+        # the creator of the group. That means that this private group is still associated
+        # with this user in some form—but we want to make sure it does not appear
+        # in these results.
```

Comment:
I realize this PR is just maintaining existing behaviour, but I don't like the inconsistency between how private and public groups behave when the user is a member. That's something I think we should try to eliminate in future to avoid confusion + code complexity.

Reply:
Part of the reason I commented so thoroughly was to expose that the logic here is fiddly. I think this absolutely warrants addressing.